File: CPP/HasCycle.cpp
/*
  Trick to know if there is a cycle is that a walker and runner will meet at some point if there is a cycle.
*/
#include <cstdlib>

struct ListNode
{
  int val;
  struct ListNode *next;
  ListNode(int x)
  {
    this->val = x;
    this->next = NULL;
  }
};

class Solution
{
public:
  bool hasCycle(struct ListNode *head)
  {
    if (head == NULL)
      return false;

    struct ListNode *walker = head;
    struct ListNode *runner = head;

    while (runner->next != NULL && runner->next->next != NULL)
    {
      walker = walker->next;
      runner = runner->next->next;

      if (walker == runner)
        return true;
    }

    return false;
  }
};

File: JS/old_solutions/HasCycle.js
/*
  Given a linked list, determine if it has a cycle in it.
  You are given pos which is a variable that tells you what index of the linked list the tail is connected to. This suggests that if the pos is less than the linked list length - 1 then there is a cycle in the list. Should do some error handling.

  The trick behind this is that a walker and runner pointer will meet if there exists a cycle. Think about it. It's a cycle so it will loop forever, it will definitely hit the same value at some point if the pacing is different.
*/
const hasCycle = function(head) {
	// A walker and runner will be the same node at some point if there is a cycle because they will circle around infinitely.
	if (head === null) return false;

	let walker = head;
	let runner = head;
	while (runner.next != null && runner.next.next != null) {
		walker = walker.next;
		runner = runner.next.next;
		if (walker === runner) return true;
	}

	return false;
};
